File: src/main/java/net/masterthought/dlanguage/stubs/types/AliasInitializerStubElementType.java
package net.masterthought.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import net.masterthought.dlanguage.psi.DLanguageAliasInitializer;
import net.masterthought.dlanguage.psi.impl.named.DLanguageAliasInitializerImpl;
import net.masterthought.dlanguage.stubs.DLanguageAliasInitializerStub;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class AliasInitializerStubElementType extends DNamedStubElementType<DLanguageAliasInitializerStub, DLanguageAliasInitializer> {
    public AliasInitializerStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DLanguageAliasInitializer createPsi(@NotNull final DLanguageAliasInitializerStub stub) {
        return new DLanguageAliasInitializerImpl(stub, this);
    }

    @NotNull
    @Override
    public DLanguageAliasInitializerStub createStub(@NotNull final DLanguageAliasInitializer psi, final StubElement parentStub) {
        return new DLanguageAliasInitializerStub(parentStub, this, psi.getName());
    }

    @Override
    public void serialize(@NotNull final DLanguageAliasInitializerStub stub, @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
    }

    @NotNull
    @Override
    public DLanguageAliasInitializerStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DLanguageAliasInitializerStub(parentStub, this, dataStream.readName());
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/AliasInitializerStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.impl.named.DLanguageAliasInitializerImpl
import io.github.intellij.dlanguage.psi.named.DLanguageAliasInitializer
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes.Companion.read
import io.github.intellij.dlanguage.stubs.DLanguageAliasInitializerStub
import java.io.IOException

class AliasInitializerStubElementType(debugName: String) :
    DNamedStubElementType<DLanguageAliasInitializerStub, DLanguageAliasInitializer>(debugName) {
    override fun createPsi(stub: DLanguageAliasInitializerStub): DLanguageAliasInitializer {
        return DLanguageAliasInitializerImpl(stub, this)
    }

    override fun createStub(
        psi: DLanguageAliasInitializer,
        parentStub: StubElement<*>
    ): DLanguageAliasInitializerStub {
        return DLanguageAliasInitializerStub(parentStub, this, psi.name, psi.attributes)
    }

    @Throws(IOException::class)
    override fun serialize(stub: DLanguageAliasInitializerStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)
        stub.attributes.write(dataStream)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>): DLanguageAliasInitializerStub {
        return DLanguageAliasInitializerStub(
            parentStub, this, dataStream.readName(),
            read(dataStream)
        )
    }
}
